File: src/main/java/me/fourteendoggo/xcore/listeners/PlayerListener.java
package me.fourteendoggo.xcore.listeners;

import me.fourteendoggo.xcore.XCore;
import me.fourteendoggo.xcore.commands.vanish.VanishManager;
import me.fourteendoggo.xcore.storage.UserManager;
import me.fourteendoggo.xcore.user.User;
import me.fourteendoggo.xcore.utils.Lang;
import me.fourteendoggo.xcore.utils.Settings;
import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.GameMode;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityPickupItemEvent;
import org.bukkit.event.entity.FoodLevelChangeEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.*;

public class PlayerListener implements Listener {
    private final XCore core;
    private final VanishManager vanishManager;
    private final UserManager userManager;

    public PlayerListener(XCore core, VanishManager vanishManager) {
        this.core = core;
        this.vanishManager = vanishManager;
        this.userManager = core.getUserManager();
    }

    @EventHandler
    public void onAsyncPreLogin(AsyncPlayerPreLoginEvent event) {
        if (event.getLoginResult() != AsyncPlayerPreLoginEvent.Result.ALLOWED) return;

        if (!userManager.loadIfAbsent(event.getUniqueId())) {
            event.disallow(AsyncPlayerPreLoginEvent.Result.KICK_OTHER, "Failed to load your data, " +
                    "please try again in a few minutes.\nIf this problem continues, contact staff.");
        }
    }

    @EventHandler
    public void onJoin(PlayerJoinEvent event) {
        Player joinedPlayer = event.getPlayer();
        // check if the player was vanished before quitting
        User user = userManager.getUser(joinedPlayer.getUniqueId());
        if (vanishManager.vanishIfFlagged(user)) {
            // notify staff a vanished player has been joined
            notifyStaff(joinedPlayer, Lang.JOIN_VANISHED, joinedPlayer.getName());
            event.setJoinMessage(null);
        } else {
            // normal player join
            event.setJoinMessage(Lang.JOIN_MESSAGE.asString(joinedPlayer.getName()));
        }
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent event) {
        Player quitPlayer = event.getPlayer();
        User user = userManager.unloadUser(quitPlayer.getUniqueId());
        // un-vanish if player was vanished, re-vanish when rejoining
        if (vanishManager.unvanish(user, true, false)) {
            // notify staff a vanished player has quit
            notifyStaff(quitPlayer, Lang.QUIT_VANISHED, quitPlayer.getName());
            event.setQuitMessage(null);
        } else {
            // normal player quit
            event.setQuitMessage(Lang.QUIT_MESSAGE.asString(quitPlayer.getName()));
        }
    }

    private void notifyStaff(Player ignoredPlayer, Lang message, Object... args) {
        for (User user : userManager) {
            Player player = user.getPlayer();
            if (player != ignoredPlayer && player.hasPermission("xcore.moderator")) {
                user.sendMessage(message, args);
            }
        }
    }

    @EventHandler(ignoreCancelled = true)
    public void onTeleport(PlayerTeleportEvent event) {
        Player player = event.getPlayer();
        if (event.getFrom().getWorld() != event.getTo().getWorld() && vanishManager.isVanished(player)) {
            allowFlight(player);
        }
    }

    @EventHandler(ignoreCancelled = true)
    public void onGameModeChange(PlayerGameModeChangeEvent event) {
        GameMode newGameMode = event.getNewGameMode();
        if (newGameMode != GameMode.ADVENTURE && newGameMode != GameMode.SURVIVAL) return;

        Player player = event.getPlayer();
        if (!vanishManager.isVanished(player)) return;

        allowFlight(player);
    }

    @EventHandler
    public void onDeath(PlayerDeathEvent event) {
        Player player = event.getEntity();
        if (!vanishManager.isVanished(player)) return;

        String deathMessage = event.getDeathMessage();
        event.setDeathMessage(null);

        // when a player got killed with /kill, we seem to have to delay the ::setAllowFlight call with 2 ticks
        int flightDelay = deathMessage != null && deathMessage.endsWith("fell out of the world") ? 2 : 1;
        allowFlight(player, flightDelay);
    }

    private void allowFlight(Player target) {
        allowFlight(target, 1);
    }

    private void allowFlight(Player target, int afterTicks) {
        boolean wasFlying = target.isFlying();
        Bukkit.getScheduler().runTaskLater(core, () -> {
            target.setAllowFlight(true);
            if (wasFlying) {
                target.setFlying(true);
            }
        }, afterTicks);
    }

    @EventHandler(ignoreCancelled = true)
    public void onFoodLevelChange(FoodLevelChangeEvent event) {
        Player player = (Player) event.getEntity();
        int newFoodLevel = event.getFoodLevel();
        if (newFoodLevel < player.getFoodLevel() && vanishManager.isVanished(player)) {
            event.setCancelled(true);
        }
    }

    @EventHandler(ignoreCancelled = true)
    public void onItemPickUp(EntityPickupItemEvent event) {
        if (!(event.getEntity() instanceof Player player)) return;
        if (Settings.getPickupItemsOnVanish()) return;
        if (!vanishManager.isVanished(player)) return;

        event.setCancelled(true);
    }

    @EventHandler(ignoreCancelled = true)
    public void onChat(AsyncPlayerChatEvent event) {
        event.setFormat(ChatColor.GRAY + event.getFormat()); // go away ugly white color
    }
}

File: src/main/java/me/fourteendoggo/xcore/listeners/PlayerListener.kt
package me.fourteendoggo.xcore.listeners

import me.fourteendoggo.xcore.XCore
import me.fourteendoggo.xcore.commands.vanish.VanishManager
import me.fourteendoggo.xcore.utils.LangKey
import me.fourteendoggo.xcore.utils.Settings
import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.GameMode
import org.bukkit.entity.Player
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityPickupItemEvent
import org.bukkit.event.entity.FoodLevelChangeEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.AsyncPlayerChatEvent
import org.bukkit.event.player.AsyncPlayerPreLoginEvent
import org.bukkit.event.player.AsyncPlayerPreLoginEvent.Result
import org.bukkit.event.player.PlayerGameModeChangeEvent
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.event.player.PlayerTeleportEvent

@Suppress("UNUSED")
class PlayerListener(private val core: XCore, private val vanishManager: VanishManager) : Listener {
    private val userManager = core.userManager

    @EventHandler
    fun onAsyncPreLogin(event: AsyncPlayerPreLoginEvent) {
        if (event.loginResult != Result.ALLOWED) return
        if (userManager.loadIfAbsent(event.uniqueId)) return

        event.disallow(Result.KICK_OTHER, PRE_LOGIN_FAILED_LOADING_DATA)
    }

    @EventHandler
    fun onJoin(event: PlayerJoinEvent) {
        val joinedPlayer = event.player
        // check if player was vanished before quitting, if so re-apply it
        val user = userManager.getUser(joinedPlayer.uniqueId)
        if (vanishManager.vanishIfFlagged(user)) {
            broadcastStaff(joinedPlayer, LangKey.JOIN_VANISHED, joinedPlayer.name)
            event.joinMessage = null
        } else {
            // normal player join
            event.joinMessage = LangKey.JOIN_MESSAGE.asString(joinedPlayer.name)
        }
    }

    @EventHandler
    fun onQuit(event: PlayerQuitEvent) {
        val quitPlayer = event.player
        val user = userManager.getUser(quitPlayer.uniqueId)
        // unvanish if player was vanished, re-vanish when rejoining
        // TODO: why?
        if (vanishManager.unvanish(user, check = true, makePersistent = false)) {
            broadcastStaff(quitPlayer, LangKey.QUIT_VANISHED, quitPlayer.name)
            event.quitMessage = null
        } else {
            // normal player quit
            event.quitMessage = LangKey.QUIT_MESSAGE.asString(quitPlayer.name)
        }
    }

    /**
     * Broadcasts a message to all staff members, ignoring [ignoredPlayer], regardless of their permissions
     */
    private fun broadcastStaff(ignoredPlayer: Player, message: LangKey, vararg args: Any) {
        for (user in userManager) {
            val player = user.player
            if (player !== ignoredPlayer && player.hasPermission("xcore.moderator")) {
                user.sendMessage(message, args)
            }
        }
    }

    @EventHandler(ignoreCancelled = true)
    fun onTeleport(event: PlayerTeleportEvent) {
        val player = event.player
        val destinationWorld = event.to?.world ?: return
        if (event.from.world != destinationWorld) return

        allowFlight(player)
    }

    @EventHandler(ignoreCancelled = true)
    fun onGameModeChange(event: PlayerGameModeChangeEvent) {
        val newGameMode = event.newGameMode
        if (newGameMode != GameMode.ADVENTURE && newGameMode != GameMode.SURVIVAL) return

        val player = event.player
        if (!vanishManager.isVanished(player)) return
        allowFlight(player)
    }

    @EventHandler
    fun onDeath(event: PlayerDeathEvent) {
        val player = event.entity
        if (!vanishManager.isVanished(player)) return

        val oldDeathMessage = event.deathMessage
        event.deathMessage = null

        // when a player got killed with /kill, we need to delay setAllowFlight() call with at least 2 ticks
        // don't ask me why
        // TODO: take locale into account
        val flightDelayTicks = if (oldDeathMessage?.endsWith("fell out of the world") == true) 2 else 1
        allowFlight(player, flightDelayTicks)
    }

    private fun allowFlight(target: Player, delayTicks: Int = 1) {
        val wasFlying = target.isFlying
        Bukkit.getScheduler().runTaskLater(core, Runnable {
            target.allowFlight = true
            if (wasFlying) target.isFlying = true
        }, delayTicks.toLong())
    }

    @EventHandler(ignoreCancelled = true)
    fun onFoodLevelChange(event: FoodLevelChangeEvent) {
        val player = event.entity as Player // FIXME: unchecked cast
        val newFoodLevel = event.foodLevel

        // ignore food level changes while being vanished
        if (newFoodLevel < player.foodLevel && vanishManager.isVanished(player)) {
            event.isCancelled = true
        }
    }

    /**
     * Cancels the pickup event if:
     * - The entity picking up something is a player
     * - The settings disallow picking something up when vanished
     * - The player is actually vanished
     */
    @EventHandler(ignoreCancelled = true)
    fun onItemPickup(event: EntityPickupItemEvent) {
        if (event.entity !is Player) return
        if (Settings.pickupItemsOnVanish) return
        if (vanishManager.isVanished(event.entity as Player)) return

        event.isCancelled = true
    }

    @EventHandler(ignoreCancelled = true)
    fun onChat(event: AsyncPlayerChatEvent) {
        // get rid of that ugly white color, like come on
        event.format = "${ChatColor.GRAY}${event.format}"
    }

    companion object {
        private const val PRE_LOGIN_FAILED_LOADING_DATA = """
            Failed to load your data, please try again in a few minutes.
            If this issue persists, contact staff.
        """
    }
}